dispatcher/service/main.py

```diff
@@ -234,7 +236,31 @@ async def main(self) -> None:
             await self.start_working()
 
             logger.info(f'Dispatcher node_id={self.node_id} running forever, or until shutdown command')
-            await self.events.exit_event.wait()
+
+            while True:
+                try:
+                    await asyncio.wait_for(self.events.exit_event.wait(), 5.0)
+
+                    # exit_event has fired, process should exit
+                    break
+                except asyncio.TimeoutError:
+                    logger.info(f'initiating broker self check for node-id {self.node_id}')
+                    for producer in self.producers:
+                        if isinstance(producer, BrokeredProducer):
+                            result = await producer.broker.get_self_check_result(self.node_id)
```

Comment:
What's actually super non-obvious is that it might not be okay to run 2 listening tasks at the same time. There are actually a lot of hazards with connection management in async code, and every I've fought this I've given up and went back to always forever only having a single task interact with psycopg.

Separate from that, this wouldn't have a timeout anyway, which is the failure scenario.